src/example-types/results.js

```diff
@@ -1,79 +1,94 @@
 let F = require('futil-js')
 let _ = require('lodash/fp')
-let Promise = require('bluebird')
 
-module.exports = {
-  result(context, search, schema, { getSchema }) {
-    let {
-      page = 1,
-      pageSize = 10,
-      sortField = '_score',
-      sortDir = 'desc',
-      populate,
-    } = context
-
-    page -= 1
-    let startRecord = page * pageSize
+let lookupFromPopulate = getSchema =>
+  F.mapIndexed((x, as) => {
+    let targetSchema = getSchema(x.schema) //|| toSingular(as), //<-- needs compromise-fp
+    if (!targetSchema)
+      throw Error(`Couldn't find schema configuration for ${x.schema}`)
+    if (!targetSchema.mongo)
+      throw Error(
+        'Populating a non mongo provider schema on a mongo provider schema is not supported'
+      )
+    let targetCollection = _.get('mongo.collection', targetSchema)
+    if (!targetCollection)
+      throw Error(
+        `The ${targetCollection} schema has a mongo configuration, but doesn't have a 'collection' property`
+      )
 
-    let sort = {
-      [sortField]: sortDir === 'asc' ? 1 : -1,
+    return {
+      $lookup: {
+        as,
+        from: targetCollection,
+        localField: x.localField, // || '_id',
+        foreignField: x.foreignField, // || context.schema, <-- needs schema lookup
+      },
     }
+  })
+
+let getStartRecord = (page, pageSize) => {
+  page -= 1
+  return page * pageSize
+}
+
+let getResultsQuery = (context, getSchema, startRecord) => {
+  let { pageSize, sortField, sortDir, populate, include } = context
+
+  let $sort = {
+    [sortField]: sortDir === 'asc' ? 1 : -1,
+  }
 
-    return Promise.all([
-      search(
-        _.reject(_.isEmpty, [
-          {
-            $sort: sort,
-          },
-          {
-            $skip: startRecord,
-          },
-          pageSize > 0 && {
-            $limit: pageSize,
-          },
-          ...F.mapIndexed((x, as) => {
-            let targetSchema = getSchema(x.schema) //|| toSingular(as), //<-- needs compromise-fp
-            if (!targetSchema)
-              throw Error(`Couldn't find schema configuration for ${x.schema}`)
-            if (!targetSchema.mongo)
-              throw Error(
-                'Populating a non mongo provider schema on a mongo provider schema is not supported'
-              )
-            let targetCollection = _.get('mongo.collection', targetSchema)
-            if (!targetCollection)
-              throw Error(
-                `The ${targetCollection} schema has a mongo configuration, but doesn't have a 'collection' property`
-              )
+  // $sort, $skip, $limit
+  let sortSkipLimit = [
+    { $sort },
+    { $skip: startRecord },
+    pageSize > 0 && {
+      $limit: pageSize,
+    },
+  ]
+  // If sort field contains a '.' move $sort, $skip, and $limit to after $lookup.
+  // Otherwise, place those first to take advantage of any indexes on that field.
+  let sortOnJoinField = /\./.test(sortField)
```

Comment:
We should explicitly check if `sortField` contains a path that is defined as part of a `$lookup` (I think it is the `as`) because it is reasonable common for fields to have a `.` (e.g. nested objects)

Reply:
Done